scalding-spark/src/main/scala/com/twitter/scalding/spark_backend/SparkWriter.scala

```diff
@@ -161,19 +168,32 @@ class SparkWriter(val sparkMode: SparkMode) extends Writer {
 
     def force[T](opt: TypedPipe[T], keyPipe: TypedPipe[T], oldState: State): (State, Action) = {
       val promise = Promise[RDD[_ <: T]]()
-      val (newState, added) = oldState.addForce[T](
+      val (newState, added, forcedRDD) = oldState.addForce[T](
         conf,
         init = keyPipe,
         opt = opt,
         promise.future,
         conf.getForceToDiskPersistMode.orElse(Some(StorageLevel.DISK_ONLY))
       )
       def action = () => {
-        // actually run
+
+        /**
+         * actually run
+         *
+         * count is a hack to force a synchronous evaluation of the RDD. There is currently only a synchronous
+         * API for un-persist but not for persist. This strategy have been used even inside spark sql itself:
+         * https://github.com/apache/spark/pull/30403/files#diff-aa467582590608ec702c1caf6208e79af5e6f3ee2cfd58a53bdf2601cd5015d6R65
+         * so it's not too crazy, but it does have a performance overhead that we will want to address in the
+         * future.
+         */
         val op = planner(opt)
-        val rddF = op.run(session)
+        val rddF = op
+          .run(session)
         promise.completeWith(rddF)
-        rddF.map(_ => ())
+        forcedRDD.map { x =>
+          x.count // any cheap RDD action to force DAG execution works here
+          ()
+        }
       }
```

Comment:
okay I currently have this hack to make the addForce method be blocking. I'm sure there is a better way but I currently can't think of one. Is this alright for now?

Reply:
I think we actually want:

```scala
forcedRdd.flatMap { x =>
  Future {
    scala.concurrent.blocking {
      x.count()
      ()
    }
  }
}
```

Because we don't want to block the mapping thread. (we should review that we are using `scala.concurrent.blocking` on blocking operations.

Reply:
We still need to address this. Otherwise this call may block a compute thread which can really harm concurrency.